Add type definitions for improved IDE/VS Code experience

aloisklink opened the pull request:
Adds `*.d.ts` files into the `types/` folder, which are auto-generated from all of the amazing JSDoc that's been made.

This allows nqm-api-tdx to work with TypeScript tools, such as the super important VS Code autocomplete/code suggestions. It will make my life (and everyone else's) much easier when programming with the TDXApi.

You can even CTRL+CLICK on one of the functions in VS Code, and it will send you directly to the function definition in `src/`, skipping the Webpack transpiled code.

In order to get this working, I had to do two potentially risky things:

- Firstly, I needed to update all the `devDependencies`, since `npm install` did not work on NPM v7, due to conflicting versions of eslint/webpack/babel. This also meant I had to update the webpack/babel config slightly, but I believe they should be roughly the same.
- Secondly, I needed to change the TDXApiError class to use the ES6 `class TDXApiError extends Error {...}` syntax.
  This also changed the JSDoc documentation slightly (TDXApiError now appears in the Classes section instead of Global), and the JSDoc is now defined with the class in the `helpers.js` file.
  Babel 7 supports transpiling this for basically every browser except for Internet Explorer 10. Internet Explorer 11 is still supported, and IE 10 is not even supported by Microsoft since 2020, so hopefully this shouldn't be an issue.

Finally, I deleted the `lib/` folder from the git repo, and instead put the build scripts in `npm run prepare`, as it's run automatically before `npm publish`, and when running `npm install` locally/from git.

## Tests

Running `npm test` (with `test/test-config.js` set to nqm1 (by the way, this repo is public, and so that sharekey for nqm1 is also public)) works (albeit very slow):

![npm test result](https://user-images.githubusercontent.com/19716675/113718353-7683ef80-96e4-11eb-8eed-775b0db96e47.png)

I also tried running `npm pack` to create a `.tgz` of this package, and ran tests with [@nqminds/aliyun-oss-bridge](https://www.npmjs.com/package/@nqminds/aliyun-oss-bridge) and I can confirm it works.

The new `lib/nqm-api-tdx.min.js` are also slightly smaller, moving from 33.4 kB to 31.3 KiB:

```console
> @nqminds/nqm-api-tdx@0.9.1 build
> webpack --env build

The babel/generator-star-spacing rule is deprecated. Please use the built in generator-star-spacing rule instead.
asset nqm-api-tdx.min.js 31.3 KiB [emitted] [minimized] (name: main) 1 related asset
runtime modules 663 bytes 3 modules
orphan modules 24.9 KiB [orphan] 1 module
built modules 109 KiB [built]
  ./src/api-tdx.js + 1 modules 109 KiB [built] [code generated]
  external "base-64" 42 bytes [built] [code generated]
  external "debug" 42 bytes [built] [code generated]
  external "@nqminds/nqm-core-utils" 42 bytes [built] [code generated]
  external "cross-fetch" 42 bytes [built] [code generated]
webpack 5.28.0 compiled successfully in 2280 ms
```

## Example Images in VS Code

![Example autocomplete with new TDX Api](https://user-images.githubusercontent.com/19716675/113713997-d7f58f80-96df-11eb-9a57-e20dd09e2df8.png)

![Example suggestions in VS Code for the TDX Api](https://user-images.githubusercontent.com/19716675/113715439-68809f80-96e1-11eb-8314-7289901490f7.png)

aloisklink committed:
Updates devDependencies so npm install works

NPM v7 fails with npm install, since there is a conflict
between eslint: ^3 and the minimum eslint version required be jest.

Unfortunately, updating eslint meant I had to update
eslint-plugin-babel, which meant I had to update webpack/babel and
etc, and therefore their configs.

aloisklink committed:
Update all (dev) dependencies to latest version

No reason to keep them on older versions, especially since
some of them have security issues/depreciation warnings.

aloisklink committed:
Confirm all tests working using jest

I was struggling to get the jest tests working,
probably because I was using `nqm-1.com` instead of a local
tdx.

I've set the maxWorkers: 1 to disable parallel tests
(it seemed to be causing issues for some tests),
and increased timeouts for all the slow tests.

aloisklink committed:
Remove lib/ from .git repo

Whenever you run `npm install github:nqminds/nqm-api-tdx`,
the npm prepare script will automatically build lib/ instead.

aloisklink committed:
Add TypeScript type definitions (*.d.ts) files

These TypeScript type definition files allows JS IDEs (e.g. VS Code)
to show autosuggestions for methods/parameters, as well as their
associated JSDoc comments.

These TypeScript type definition files are completely seperate from
the files generated by Webpack, and are stored in the types/
directory.

The only JavaScript code I changed is how TDXApiError is defined,
I had to change it to use ES6: class TDXApiError extends Error {...}
syntax, as TypeScript wasn't happy with the old format.

Luckily, Babel 7 supports the `extend Error` syntax now, on all
browsers except IE 10 (IE 11 still supported). Otherwise, the
behaviour should be identical.

I also slightly modified the JSDoc process by making documenting
TDXApiError as a class, so it syncs better with VS Code/TypeScript

iendjinn commented on CHANGELOG.md:
```diff
@@ -6,6 +6,23 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+  - base-64: ^1.0.0
+  - cross-fetch: ^3.1.2
+  - debug: ^4.3.1
+- Update devDepdencies to work with NPM v7 `npm install`:
```
I assume you've tested this with NPM < 7 as well? Think most people don't upgrade _that_ often.

aloisklink replied:
NPM < 4 will not work, since `npm run prepare` only became a feature in version 4.0.0, but the last NPMv3 release was in 2016, so I think we're probably okay on that side.

NPM version 6 works, but prints out a warning that it can't fully understand the `package-lock.json` file (it's backwards compatible with NPMv5/v6). The only bad thing is that doing if people are constantly developing with NPM v6/v7, the lockfile will keep adding/deleting the extra `v2` data that NPMv7 injects. But it's would just make our `git diffs` look bigger, both lock-files are compatible with each version of NPM.

```
npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!
```

iendjinn commented on package.json:
```diff
@@ -20,28 +29,30 @@
+    "jest": "^27.0.0-next.6",
+    "jsdoc": "^3.6.6",
+    "lodash": "^4.17.21",
+    "npm-run-all": "^4.1.5",
```
A ~100kb library to do this? :disappointed_relieved: 

Next thing you'll be using lodash map...

aloisklink replied:
And it has 9 dependencies as well! I mainly didn't want to write `npm run build && npm run build:dev` since I was worried about the `&&` on non-Linux platforms, and hard-coding `npm` in case someone wants to use `yarn`.

However, it looks like https://www.npmjs.com/package/yarn-or-npm is tiny, and has much less dependencies, and `&&` is supported by both Unix and Windows, and I doubt there are any other OSes anybody is using Node with, so I could switch to using https://www.npmjs.com/package/yarn-or-npm instead.

iendjinn replied:
Supporting windows is undesirable :)

aloisklink replied:
Awesome, I've switched to `yarn-or-npm`, so we still have both Windows and Yarn supported (and removed like 2000 lines from our `package-lock.json` file).

Supporting Windows isn't too much trouble though, just for the off chance that one of the Windows users (@nallott or Uni of Soton) decides to install via GitHub, not npm. At least we don't need to worry about esoteric platforms like the [Unisys 1100/2200](https://begriffs.com/posts/2018-11-15-c-portability.html) where one `char` is 9 bits, and has a +0 and -0.

iendjinn reviewed (approved):
LGTM.

Thanks for updating the webpack conf, that must have been a bit of a pain.

aloisklink committed:
Replace npm-run-all with smaller yarn-or-npm

npm-run-all is a massive 100KiB package with
9 dependencies.

yarn-or-npm on the other hand is only 5.33 KiB large and only
has 2 direct dependencies.

The package-lock.json file is quite a bit smaller as well.

aloisklink commented:
Updating the webpack config wasn't too bad actually, the new webpack/babel version is still annoying, but much easier to use.

I considering changing the library export to a dual export CommonJS/ESM export, but I didn't want to damage any backwards comparability with the old UMD library format, so I kept it as similar to the previous format as I could.

TobyEalden reviewed (changes requested):
If you need to publish this before the next TDX release (probably 2-3 weeks away) can you bump the version to 0.10

aloisklink commented:
Hi @TobyEalden, this is mainly just a quality-of-life/documentation improvement, it's not really adding any other features to the TDX API (other than fixing `npm install --development` on NPMv7).

Feel free to release this with the next TDX release. I've been waiting since I was an intern 3 years ago for this, I can wait a few more weeks :smile:
